### PR TITLE
Failsafe - Fixes for RX loss detection.

### PR DESCRIPTION
**Pull request by hydra**
Large code re-organization which separates some key tasks in the rx code.

Tests need to be updated before merging.

Tested with X8R in SBus and PWM, and Spek Sat.

See #1190 and #1207

### CODE REVIEW

File: src/main/drivers/pwm_rx.c [hunk 1]
```diff
@@ -227,6 +227,17 @@ static void ppmEdgeCallback(timerCCHandlerRec_t* cbRec, captureCompare_t capture
 
 #define MAX_MISSED_PWM_EVENTS 10
 
+bool isPWMDataBeingReceived(void)
+{
+    int channel;
+    for (channel = 0; channel < PWM_PORTS_OR_PPM_CAPTURE_COUNT; channel++) {
+        if (captures[channel] != PPM_RCVR_TIMEOUT) {
```
Comment (ledvinap) [inline src/main/drivers/pwm_rx.c]: Will this work if one channel is used for RSSI?

File: src/main/rx/rx.c [hunk 2]
```diff
@@ -228,24 +238,37 @@ bool rxIsReceivingSignal(void)
...
+bool rxAreFlightChannelsValid(void)
+{
+    return rxFlightChannelsValid;
+}
 static bool isRxDataDriven(void) {
     return !(feature(FEATURE_RX_PARALLEL_PWM | FEATURE_RX_PPM));
 }
 
-void updateRx(uint32_t currentTime)
+static void resetRxSignalTimeoutIfNeeded(uint32_t currentTime)
```
Comment (ledvinap) [inline src/main/rx/rx.c]: `testRxSignalTimeout`? or maybe `processRxSignalTimeout`

Timeout is not reset in this function

File: src/main/rx/rx.c [hunk 3]
```diff
@@ -285,8 +295,15 @@ void updateRx(uint32_t currentTime)
             needRxSignalBefore = currentTime + DELAY_10_HZ;
             resetPPMDataReceivedState();
         }
-        shouldCheckPulse = rxSignalReceived;
     }
+
+    if (feature(FEATURE_RX_PARALLEL_PWM)) {
```
Comment (ledvinap) [inline src/main/rx/rx.c]: feature(FEATURE_RX_PARALLEL_PWM) is tested twice (https://github.com/cleanflight/cleanflight/pull/1224/files#diff-7cb1b466189980bc90ef231c7828fb58R288)

**Comment by ProDrone** [comment]
@hydra I commented here: https://github.com/cleanflight/cleanflight/issues/1207

File: src/main/rx/rx.c [hunk 4]
```diff
@@ -228,24 +238,37 @@ bool rxIsReceivingSignal(void)
     return rxSignalReceived;
 }
 
+bool rxAreFlightChannelsValid(void)
```
Comment (ledvinap) [inline src/main/rx/rx.c]: Logical operation is not apparent from name - this test if at least one channel is valid (is this intended? Disconnected wire? )

File: src/main/drivers/pwm_rx.c (hunk 1 above)
Comment (hydra) [inline src/main/drivers/pwm_rx.c]: I'll double check that, thanks for the reminder.

File: src/main/rx/rx.c (hunk 4 above)
Comment (hydra) [inline src/main/rx/rx.c]: Yeah, i was finding it hard to name it well.  Yes, should be false if any of the 4 flight channels are invalid for whayever reason, including disconnected PWM cable or out of range stick.

File: src/main/rx/rx.c (hunk 2 above)
Comment (hydra) [inline src/main/rx/rx.c]: I think i renamed that twice and moved a bunch of code around, probably needs renaming again, yes.

File: src/main/drivers/pwm_rx.c (hunk 1 above)
Comment (hydra) [inline src/main/drivers/pwm_rx.c]: The answer is yes.  There should always be a pulse in the range 1000-2000 for PWM RSSI.

If some receivers don't honour this we can find out which ones and address this later.

```
    // Range of rawPwmRssi is [1000;2000]. rssi should be in [0;1023];
    rssi = (uint16_t)((constrain(pwmRssi - 1000, 0, 1000) / 1000.0f) * 1023.0f);
```

File: src/main/rx/rx.c (hunk 3 above)
Comment (hydra) [inline src/main/rx/rx.c]: fixed.

File: src/main/drivers/pwm_rx.c (hunk 1 above)
Comment (ledvinap) [inline src/main/drivers/pwm_rx.c]: But this means that with RSSI failsafe won't be triggered - RSSI channel will be always OK

**Pull request merged**